BJS2-77786

SAM19991 opened the pull request:
Карьера

SAM19991 committed:
BJS2-77786
Карьера

kashkademan commented on src/main/java/school/faang/user_service/controller/career/CareerController.java:
```diff
@@ -0,0 +1,28 @@
+public class CareerController {
+    private final CareerService careerService;
+
+    @Autowired
```
мне кажется это не обязательно писать. оно само заварится или нет?

kashkademan commented on src/main/java/school/faang/user_service/controller/career/CareerController.java:
```diff
@@ -0,0 +1,28 @@
+import school.faang.user_service.service.career.CareerService;
+
+@Controller
+public class CareerController {
```
нужно добавить мне кажется @RequestMapping() анотацию, чтобы понимать по какому адресу запрос отправлять.

kashkademan commented on src/main/java/school/faang/user_service/controller/career/CareerController.java:
```diff
@@ -0,0 +1,28 @@
+        this.careerService = careerService;
+    }
+
+    public CareerDto addCareer(long userId, CareerDto careerDto) {
```
и здесь перед методом @PostMapping("/somPath") потому что метод записывает что то.

kashkademan replied:
А оно так заработает? не надо писать сам путь? когда будешь слать запрос то к методу должен работать так например. Я так делал. но если у тебя работает, то можешь оставить Миша проверит :)

вызов сюда:
localhost:8080/career/addCareer

берется из аннотаций контроллера путь:
@RequestMapping("/career") перед самим классом
@PostMapping("/addCareer") перед методом
@RequestParam перед аргументами метода

kashkademan commented on src/main/java/school/faang/user_service/controller/career/CareerController.java:
```diff
@@ -0,0 +1,28 @@
+        return careerService.addCareer(userId, careerDto);
+    }
+
+    public CareerDto updateCareer(long userId, CareerDto careerDto) {
```
@PostMappint("") наверное должен быть.

kashkademan replied:
Поправить (или нет) с комментариями выше.

kashkademan commented on src/main/java/school/faang/user_service/controller/career/CareerController.java:
```diff
@@ -0,0 +1,28 @@
+        return careerService.updateCareer(userId, careerDto);
+    }
+
+    public CareerDto getById(long careerId) {
```
@GetMapping("") добавить.

kashkademan replied:
Поправить (или нет) с комментариями выше.

kashkademan commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,73 @@
+        if (!careerDto.getFrom().isBefore(LocalDate.now())) {
+            throw new DataValidationException("Career date can not be in the future");
+        }
+        User user = userRepository.findById(userId).orElseThrow(() -> new RuntimeException("User not found"));
```
а не должны мы тут свое исключение использовать по задаче?

kashkademan commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,73 @@
+    private final UserRepository userRepository;
+    private final CareerMapper careerMapper;
+
+    public CareerService(UserRepository userRepository,
```
лучше наверное добавить через лобок аннотация @RequiredArgsConstructor и руками не писать.

kashkademan commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,73 @@
+        }
+        User user = userRepository.findById(userId).orElseThrow(() -> new RuntimeException("User not found"));
+        Career partial = careerMapper.toCareer(careerDto);
+        Career career = Career.builder()
```
а разве надо тут его собирать заново ,он же сам по идее уже скорвертнется выше. toCareer() метод его уже превратит в Career, разве нет?

kashkademan reviewed (changes requested):
Написал комменты. А тестировал что написано, оно работает?))

SAM19991 committed:
Заменил Joda на java.time.LocalDate.

Добавил @Setter в Career для удобного обновления полей.

CareerMapper — маппит только даты.

Исправил методы в классе CareerService.

Перевёл контроллер в @RestController с @PostMapping, @PutMapping и @GetMapping.

Проставил в каждом классе где необходимо @RequiredArgsConstructor.

kashkademan commented on src/main/java/school/faang/user_service/dto/CareerDto.java:
```diff
@@ -1,13 +1,11 @@
+@Data
 public class CareerDto {
-    private long id;
+    private Long id;
```
а чего поменял на класс вместо примитивного объекта?

kashkademan commented on src/main/java/school/faang/user_service/entity/Career.java:
я бы энтити не правил, в задаче не надо и мы такое не проходили :) должно через мэппинг думаю собираться. Но можно оставить, Миша если что подтверди, пожалуйста.

SAM19991 committed:
Добавил controller/CareerController как @RestController с эндпоинтами
  • POST /users/{userId}/career
  • PUT  /users/{userId}/career
  • GET  /career/{careerId}

kashkademan commented:
Вроде выглядит норм. С именами временных можно еще подумать, заметил что не всегда понятно где dto где entity. А так ок с моей стороны.

miked4king commented on src/main/java/school/faang/user_service/controller/career/CareerController.java:
```diff
@@ -0,0 +1,35 @@
+@Controller
+@RequiredArgsConstructor
+public class CareerController {
+    private final CareerService careerService;
```
в задании у вас это не описано, но тут ты жестко завязываешься на конкретной реализации (нарушение последней буквы в solid)
хорошей практикой является использование интерфейсов, сделай CareerService интерфейсом, сделай имплементацию, обычно называют название интерфейса + Impl (CareerServiceImpl), тут у тебя все останется без изменений, спринг под капотом увидит твою единственную реализацию и будет использовать ее

miked4king commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,56 @@
+    public CareerDto updateCareer(long userId, CareerDto careerDto) {
+        if (!careerDto.getFrom().isBefore(LocalDate.now())) {
+            throw new DataValidationException("Career date can not be in the future");
+        }
```
вроде аналогичная проверка в методе выше, может вынесем в отдельный метод?

miked4king commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,56 @@
+    public CareerDto getById(long careerId) {
+        Career career = careerRepository.findById(careerId).orElseThrow(() ->
+                new RuntimeException("Career not found"));
+        return careerMapper.toCareerDto(career);
```
можем чуть элегантнее записать
return careerRepository.findById(careerId)
.map(careerMapper::toCareerDto)
.orElseThrow(() -> new RuntimeException("Career not found"));

miked4king commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,56 @@
+                new DataValidationException("User not found"));
+        Career career = careerMapper.toCareer(careerDto);
+        career.setUser(user);
+        Career result = careerRepository.save(career);
```
давай подумаем над названием переменной

miked4king commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,56 @@
+        if (!careerToUpdate.getUser().getId().equals(userId)) {
+            throw new DataValidationException("Career user id not match");
+        }
+        Career updated = careerMapper.toCareer(careerDto);
```
тоже подумай над названием переменной updated и result 
по названию переменной должно быть понятно что она хранит

SAM19991 committed:
Moved CareerService to interface and created CareerServiceImpl with implementation

Refactor CareerServiceImpl: extract date validation, rename vars, simplify getById.

SAM19991 committed:
Moved CareerService to interface and created CareerServiceImpl with implementation

Refactor CareerServiceImpl: extract date validation, rename vars, simplify getById.

miked4king reviewed (changes requested):
Все четко, принимаю работу.
Только сейчас есть конфликт с вашей мастер веткой, тебе придется предварительно его разрешить.
для этого тебе потребуется в idea подтянуть (pull) изменения из triton-master-stream10, разрешить конфликт в классе DataValidationException, и опять запушить.
после этого сможем смержить)

SAM19991 committed:
Covered CareerService with tests: addCareer, updateCareer, getById

SAM19991 committed:
77786

SAM19991 committed:
Merge branch 'triton-master-stream10' into develop/BJS2-77786
